test/main_test.cpp

```diff
@@ -0,0 +1,218 @@
+      EXPECT(frameBefore.creep_map == frameAfter.creep_map);
+      EXPECT(frameBefore.bullets == frameAfter.bullets);
+      EXPECT(matchingResources);
+      EXPECT(matchingActions);
+      EXPECT(matchingUnits);
+    }
+  },
+  
+  lest_CASE("A TorchCraft FrameDiff is invariant through serialization") {
+    SETUP("Create & serialize a FrameDiff") {
+      torchcraft::replayer::FrameDiff diffBefore, diffAfter;
+      diffBefore.reward = 1.1;
+      diffBefore.is_terminal = ! diffBefore.is_terminal;
+      diffBefore.pids = {1, 2, 3};
+      diffBefore.creep_map = {{11, 12}, {21, 22}};
+      diffBefore.bullets = {{11, 12, 13}, {21, 22, 23}};
+      diffBefore.resources = {
+        {11, {12, 13, 14, 15, 16, 17, 18}},
+        {21, {22, 23, 24, 25, 26, 27, 28}}};
+      diffBefore.actions = {
+        {100, {{{111, 112}, 113, 114}, {{121, 122}, 123, 124}}},
+        {200, {{{211, 212}, 213, 214}, {{221, 222}, 223, 224}}}};
+      diffBefore.units = {
+        {{
+          100, {121, 122}, {131, 132}, {141, 142}, {151, 152},
+          161, 162.5, 163.5, 164
+        }, {
+          200, {221, 222}, {231, 232}, {241, 242}, {251, 252},
+          261, 262.5, 263.5, 264}},
+        {{
+          300, {321, 322}, {331, 332}, {341, 342}, {351, 352},
+          361, 362.5, 363.5, 364
+        }, {
+          400, {421, 422}, {431, 432}, {441, 442}, {451, 452},
+          461, 462.5, 463.5, 464}}};
+          
+      std::stringstream buffer;
+      buffer << diffBefore;
+      buffer >> diffAfter;
+      
+      auto sortedCreepBefore = sortMap(diffBefore.creep_map);
+      auto sortedCreepAfter = sortMap(diffAfter.creep_map);
+      auto sortedResourcesBefore = sortMap(diffBefore.resources);
+      auto sortedResourcesAfter = sortMap(diffAfter.resources);
+      auto sortedActionsBefore = sortMap(diffBefore.actions);
+      auto sortedActionsAfter = sortMap(diffAfter.actions);
+
+      // Lest tries to fancy-print EXPECT arguments.
+      // But it but doesn't know how to print these complex objects.
+      // So we do the comparison here and feed Lest the results.
+      auto matchingCreep = sortedCreepBefore == sortedCreepAfter;
+      auto matchingResources = sortedResourcesBefore == sortedResourcesAfter;
+      auto matchingActions = sortedActionsBefore == sortedActionsAfter;
+      auto matchingUnits = diffBefore.units == diffAfter.units;
+      
+      EXPECT(diffBefore.reward == diffAfter.reward);
+      EXPECT(diffBefore.is_terminal == diffAfter.is_terminal);
+      EXPECT(diffBefore.pids == diffAfter.pids);
+      EXPECT(matchingCreep);
+      EXPECT(matchingResources);
+      EXPECT(matchingActions);
+      EXPECT(matchingUnits);      
+    }        
+  }
+};
+
+} // namespace torchcraft
+} // namespace replayer
+
+// For output, run either as:
+// > ctest test --output-on-failure
+// > make ARGS='-VV' test
+int main(int argc, char* argv[]) {
+  std::cout << "Running tests with console output" << std::endl;
```

Comment:
Nit: output logging/non-data to cerr maybe

Reply:
How come?

Reply:
Just generally, it's preferred to have logging output sent to stderr.